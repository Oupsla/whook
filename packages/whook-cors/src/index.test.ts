import { wrapHandlerWithCORS, optionsWithCORS, augmentAPIWithCORS } from '.';
import { handler } from 'knifecycle';
import { OpenAPIV3 } from 'openapi-types';

describe('wrapHandlerWithCORS', () => {
  const CORS = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET,POST,PUT,DELETE,OPTIONS',
    'Access-Control-Allow-Headers': [
      'Accept',
      'Accept-Encoding',
      'Accept-Language',
      'Referrer',
      'Content-Type',
      'Content-Encoding',
      'Authorization',
      'Keep-Alive',
      'User-Agent',
    ].join(','),
    Vary: 'Origin',
  };

  it('should work', async () => {
    const wrappedOptionsWithCORS = wrapHandlerWithCORS<any, any>(
      optionsWithCORS,
    );
    const wrappedHandler = await wrappedOptionsWithCORS({
      CORS,
    });
    const response = await wrappedHandler();

    expect({
      response,
    }).toMatchSnapshot();
  });

  it('should add CORS to errors', async () => {
    const wrappedGetError = wrapHandlerWithCORS<any, any>(
      handler(
        async function getError() {
          throw new Error();
        },
        'getError',
        [],
      ),
    );
    const wrappedHandler = await wrappedGetError({
      CORS,
    });

    try {
      await wrappedHandler();
      throw new Error('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        headers: err.headers,
      }).toMatchSnapshot();
    }
  });
});

describe('augmentAPIWithCORS()', () => {
  it('should work', async () => {
    expect(
      await augmentAPIWithCORS({
        openapi: '3.0.2',
        info: {
          version: '1.0.0',
          title: 'Sample OpenAPI',
          description: 'A sample OpenAPI file for testing purpose.',
        },
        components: {
          securitySchemes: {
            oAuth2: {
              type: 'oauth2',
              flows: {},
            },
          },
        },
        paths: {
          '/ping': {
            options: {
              operationId: 'optionsPing',
              summary: 'Provides ping options.',
              responses: {
                '200': {
                  description: 'Ping options',
                },
              },
            },
            get: {
              operationId: 'getPing',
              summary: "Checks API's availability.",
              responses: {
                '200': {
                  description: 'Pong',
                },
              },
            },
          },
          '/users/{userid}': {
            head: {
              operationId: 'getUser',
              summary: 'Return a user.',
              security: [
                {
                  oAuth2: ['user'],
                },
              ],
              parameters: [
                {
                  in: 'path',
                  name: 'userId',
                  required: true,
                  schema: {
                    type: 'number',
                  },
                },
                {
                  in: 'query',
                  name: 'full',
                  required: true,
                  schema: {
                    type: 'boolean',
                  },
                },
                {
                  in: 'query',
                  name: 'retry',
                  required: false,
                  schema: {
                    type: 'boolean',
                  },
                },
              ],
              responses: {
                '200': {
                  description: 'The user',
                },
              },
            },
          },
          '/crons/tokens': {
            post: {
              operationId: 'ping',
              'x-whook': {
                type: 'cron',
              },
              summary: "Checks API's availability.",
              responses: {
                '200': {
                  description: 'Pong',
                },
              },
            } as OpenAPIV3.OperationObject,
          },
        },
      }),
    ).toMatchSnapshot();
  });
});
