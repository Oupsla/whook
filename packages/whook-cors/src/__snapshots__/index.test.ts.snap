// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`augmentAPIWithCORS() should work 1`] = `
Object {
  "components": Object {
    "securitySchemes": Object {
      "oAuth2": Object {
        "flows": Object {},
        "type": "oauth2",
      },
    },
  },
  "info": Object {
    "description": "A sample OpenAPI file for testing purpose.",
    "title": "Sample OpenAPI",
    "version": "1.0.0",
  },
  "openapi": "3.0.2",
  "paths": Object {
    "/crons/tokens": Object {
      "post": Object {
        "operationId": "ping",
        "responses": Object {
          "200": Object {
            "description": "Pong",
          },
        },
        "summary": "Checks API's availability.",
        "x-whook": Object {
          "type": "cron",
        },
      },
    },
    "/ping": Object {
      "get": Object {
        "operationId": "getPing",
        "responses": Object {
          "200": Object {
            "description": "Pong",
          },
        },
        "summary": "Checks API's availability.",
      },
      "options": Object {
        "operationId": "optionsPing",
        "responses": Object {
          "200": Object {
            "description": "Ping options",
          },
        },
        "summary": "Provides ping options.",
      },
    },
    "/users/{userid}": Object {
      "head": Object {
        "operationId": "getUser",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "required": true,
            "schema": Object {
              "type": "number",
            },
          },
          Object {
            "in": "query",
            "name": "full",
            "required": true,
            "schema": Object {
              "type": "boolean",
            },
          },
          Object {
            "in": "query",
            "name": "retry",
            "required": false,
            "schema": Object {
              "type": "boolean",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "The user",
          },
        },
        "security": Array [
          Object {
            "oAuth2": Array [
              "user",
            ],
          },
        ],
        "summary": "Return a user.",
      },
      "options": Object {
        "operationId": "optionsWithCORS",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "userId",
            "required": true,
            "schema": Object {
              "type": "number",
            },
          },
          Object {
            "in": "query",
            "name": "full",
            "required": false,
            "schema": Object {
              "type": "boolean",
            },
          },
          Object {
            "in": "query",
            "name": "retry",
            "required": false,
            "schema": Object {
              "type": "boolean",
            },
          },
          Object {
            "in": "query",
            "name": "access_token",
            "required": false,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "description": "CORS sent.",
          },
        },
        "summary": "Enable OPTIONS for CORS",
        "tags": Array [
          "CORS",
        ],
        "x-whook": Object {
          "private": true,
          "sourceOperationId": "getUser",
          "suffix": "CORS",
          "type": "http",
        },
      },
    },
  },
}
`;

exports[`wrapHandlerWithCORS should add CORS to errors 1`] = `
Object {
  "headers": Object {
    "Access-Control-Allow-Headers": "Accept,Accept-Encoding,Accept-Language,Referrer,Content-Type,Content-Encoding,Authorization,Keep-Alive,User-Agent",
    "Access-Control-Allow-Methods": "GET,POST,PUT,DELETE,OPTIONS",
    "Access-Control-Allow-Origin": "*",
    "Vary": "Origin",
  },
}
`;

exports[`wrapHandlerWithCORS should work 1`] = `
Object {
  "response": Object {
    "headers": Object {
      "Access-Control-Allow-Headers": "Accept,Accept-Encoding,Accept-Language,Referrer,Content-Type,Content-Encoding,Authorization,Keep-Alive,User-Agent",
      "Access-Control-Allow-Methods": "GET,POST,PUT,DELETE,OPTIONS",
      "Access-Control-Allow-Origin": "*",
      "Vary": "Origin",
    },
    "status": 200,
  },
}
`;
